### PR TITLE
mockstore: introducing embedded unistore

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve?
Integrating unistore into TiDB, so we don't need to maintain a storage backend for testing.

The unistore backend has much better performance, so you can start a single node TiDB by set `-store=unistore` to check your performance optimizations.

### What is changed and how it works?
The main work is done in https://github.com/ngaut/unistore/pull/382. TiDB side just needs to wrap the tools provided by unistore to implement the test toolkit's interface.

In order to avoid the impact of the introduced bug on CI, I temporarily retained mocktikv and provided configuration to switch.

### Check List 

Tests 
- Unit test
- Integration test

### Release note 
* No release note

### CODE REVIEW

File: server/http_handler_test.go [hunk 1]
```diff
@@ -443,7 +441,7 @@ func (ts *HTTPHandlerTestSuite) TestGetTableMVCC(c *C) {
 	info := data.Value.Info
 	c.Assert(info, NotNil)
 	c.Assert(len(info.Writes), Greater, 0)
-	startTs := info.Writes[2].StartTs
```
Comment (bobotu) [inline server/http_handler_test.go:446]: `info.Writes` returned by mocktikv is
```
type:Lock start_ts:416637082071465985 commit_ts:416637082071728128 
type:Lock start_ts:416637082066485249 commit_ts:416637082066485250 
start_ts:416637082062553091 commit_ts:416637082063077376 short_value:"\200\000\001\000\000\000\002\001\0002" 
start_ts:416637082062553089 commit_ts:416637082062553090 short_value:"\200\000\001\000\000\000\002\001\0001"
```

by unistore is
```
start_ts:416637064772321280 commit_ts:416637064772583424 short_value:"\200\000\001\000\000\000\002\001\0002" 
start_ts:416637064772059142 commit_ts:416637064772059143 short_value:"\200\000\001\000\000\000\002\001\0001"
```

Is this a bug in unistore @cfzjywxk

**Comment by ngaut** [comment]
Good job, looking forward to seeing amazing performance improvements.

File: store/mockstore/mockstore.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+// StoreType is the type of backend mock storage.
+type StoreType uint8
+
+const (
+	// MockTiKV is the mock storage based on goleveldb.
+	MockTiKV StoreType = iota
+	// EmbedUnistore is the mock storage based on unistore.
+	EmbedUnistore
+
+	defaultStoreType = EmbedUnistore
```
Comment (bobotu) [inline store/mockstore/mockstore.go:83]: I think it is better to use `MockTiKV` as the default storage backend, and add a new CI task to use unistore (how to achieve this? use env variables?).
If the new CI task is stable enough, we can remove the mocktikv when the next time we need to implement features on mocktikv.

File: store/mockstore/unistore/mock.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		persistent = false
+	}
+
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return nil, nil, nil, err
+	}
+
+	conf := usconf.DefaultConf
+	conf.Engine.DBPath = path
+	conf.Server.Raft = false
```
Comment (coocood) [inline store/mockstore/unistore/mock.go:43]: Some other parameters may need to change depends on the path.
If path is "", it is run in unit test, we can set some config for this case, the memtable can be smaller, 1 compaction worker, do not sync write, etc.

File: server/http_handler_test.go (hunk 1 above)
Comment (cfzjywxk) [reply server/http_handler_test.go:446]: Seems the `Op_lock` is not returned in `unistore`, could change the case  temporarily  to
```
startTs := info.Writes[0].StartTs
```
to let it pass @bobotu

File: server/http_handler_test.go [hunk 4]
```diff
@@ -441,7 +442,17 @@ func (ts *HTTPHandlerTestSuite) TestGetTableMVCC(c *C) {
 	info := data.Value.Info
 	c.Assert(info, NotNil)
 	c.Assert(len(info.Writes), Greater, 0)
-	startTs := info.Writes[0].StartTs
+
+	// Unistore will not return Op_Lock.
```
Comment (ngaut) [inline server/http_handler_test.go:446]: Please add a todo here.

**Comment by bobotu** [comment]
The default mockstore for testing is still the `mocktikv`, I'll change it after some bug in unistore is stable enough to run CI cases.

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
Could you file an issue to reflect the future plan?

> The default mockstore for testing is still the `mocktikv`, I'll change it after some bug in unistore is stable enough to run CI cases.

File: tidb-server/main.go [hunk 5]
```diff
@@ -286,7 +286,9 @@ func registerStores() {
 	err := kvstore.Register("tikv", tikv.Driver{})
 	terror.MustNil(err)
 	tikv.NewGCHandlerFunc = gcworker.NewGCWorker
-	err = kvstore.Register("mocktikv", mockstore.MockDriver{})
+	err = kvstore.Register("mocktikv", mockstore.MockTiKVDriver{})
+	terror.MustNil(err)
+	err = kvstore.Register("unistore", mockstore.EmbedUnistoreDriver{})
```
Comment (lysu) [inline tidb-server/main.go:291]: maybe we also need to add "unistore" to main.go:119, https://github.com/pingcap/tidb/pull/17156/files#diff-af2c9279ed41f7023f9b6b73cba2d18fR119

**Comment by bobotu** [comment]
> 
> 
> Could you file an issue to reflect the future plan?
> 
> > The default mockstore for testing is still the `mocktikv`, I'll change it after some bug in unistore is stable enough to run CI cases.

https://github.com/pingcap/tidb/issues/17266

**Pull request merged**